zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+                    Error::<T>::Unexpected
+                );
+                ensure!(amount_in <= pool.calculate_max_amount_in(), Error::<T>::NumericalLimits);
+                // Instead of first executing a swap with `(n-1)` transfers from the pool account to
+                // `who` and then selling complete sets, we prevent `(n-1)` storage reads: 1)
+                // Transfer `amount_in` units of `asset_in` to the pool account, 2) sell
+                // `amount_out` complete sets using the pool account, 3) transfer
+                // `amount_out_minus_fees` units of collateral to `who`. The fees automatically end
+                // up in the pool.
+                let amount_out = pool.calculate_swap_amount_out_for_sell(asset_in, amount_in)?;
+                // Beware! This transfer happen _after_ calculating `amount_out`:
+                T::AssetManager::transfer(asset_in, &who, &pool.account_id, amount_in)?;
+                T::CompleteSetOperations::sell_complete_set(
+                    pool.account_id.clone(),
+                    market_id,
+                    amount_out,
+                )?;
+                let FeeDistribution {
+                    remaining: amount_out_minus_fees,
+                    swap_fees: swap_fee_amount,
+                    external_fees: external_fee_amount,
+                } = Self::distribute_fees(market_id, pool, amount_out)?;
+                ensure!(amount_out_minus_fees >= min_amount_out, Error::<T>::AmountOutBelowMin);
+                T::AssetManager::transfer(
+                    pool.collateral,
+                    &pool.account_id,
+                    &who,
+                    amount_out_minus_fees,
+                )?;
+                for (&asset, balance) in pool.reserves.iter_mut() {
+                    if asset == asset_in {
+                        *balance = balance.checked_add(&amount_in).ok_or(Error::<T>::MathError)?;
+                    }
+                    *balance = balance.checked_sub(&amount_out).ok_or(Error::<T>::MathError)?;
+                }
+                let new_price = pool.calculate_spot_price(asset_in)?;
+                ensure!(
+                    new_price >= MIN_SPOT_PRICE.saturated_into(),
+                    Error::<T>::SpotPriceBelowMin
+                );
+                Self::deposit_event(Event::<T>::SellExecuted {
+                    who: who.clone(),
+                    market_id,
+                    asset_in,
+                    amount_in,
+                    amount_out,
+                    swap_fee_amount,
+                    external_fee_amount,
+                });
+                Ok(())
+            })
+        }
+
+        #[require_transactional]
+        fn do_join(
+            who: T::AccountId,
+            market_id: MarketIdOf<T>,
+            pool_shares_amount: BalanceOf<T>,
+            max_amounts_in: Vec<BalanceOf<T>>,
+        ) -> DispatchResult {
```

Comment:
If I understand correctly, the function assumes that the caller already bought a sufficient number of full sets. However, the parameter that determines the amounts of outcome tokens to be transferred into the pool is `pool_shares_amount`, thus the caller has to know how many `pool_shares_amounts` are acceptable at most to satisfy the boundaries specified within `max_amounts_in`. That seems like a terrible UX. As a user I am thinking about "I want to add this amounts of liquidity to the pool" instead of "I want to receive that amount of pool shares". It seems like this just introduced additional work for the front end to always calculate the `pool_shares_amount`, as it will probably just offer the user to "deposit X b That's also how other pools usually work.

Reply:
Why is it important for the user at all to know how many pool shares they get in total?

Reply:
> That seems like a terrible UX.

I don't think it's that bad. First of all, I'm copying the old API; `pool_join` from zrml-swaps had the same interface. In the end, both approaches are "equivalent" in the sense that you can quickly translate `pool_shares_amount` and `max_amounts_in` into what you're suggesting and vice versa. I think we frontend can compensate this well.

> Why is it important for the user at all to know how many pool shares they get in total?

It's (almost) completely irrelevant. If you `join` and then immediately `exit`, you should get out exactly what you put in (modulo exit fee and rounding). We can give that guarantee regardless of how many pool shares you receive. And if you wait at least a block, then there's no telling what you're going to get out of the pool, regardless of how many shares you have. Who knows what trades the informants will take, right?

So I think we may have to rethink this API a little bit, but I don't think I can do it in the next hour.